README
/*
  Name:         Zeta Dva
  Description:  Amateur level chess engine
  Author:       Srdja Matovic <s.matovic@app26.de>
  Created at:   2011-01-15
  Updated at:   2016-07-13
  License:      GPL >= v2

  Copyright (C) 2011-2016 Srdja Matovic

  Zeta Dva is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 2 of the License, or
  (at your option) any later version.

  Zeta Dva is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
*/

### Binaries ###
Linux 64 bit and Windows 32/64 bit binaries are placed in bin folder. 

### Play ### 
To play against the engine use an CECP v2 protocol capable chess GUI
like Arena, Winboard or Xboard.

Alternatively you can use xboard commmands directly on commmand line,
e.g.:
new            // init new game from start position
level 40 4 0   // set time control to 4 minutes each 40 moves
go             // let engine play site to move
usermove d7d5  // let engine apply usermove and start thinking

Non-Xboard commands:
perft          // perform a performance test, depth set by sd command
selftest       // run an internal selftest
help           // print usage hints
log            // toggle log flag

Not supported xboard commands:
analyze        // enter analyze mode
undo/remove    // take back last moves
?              // move now
draw           // handle draw offers
hard/easy      // turn on/off pondering
hint           // give user a hint move
bk             // book lines
pause/resume   // pause the engine

### Opening Book ###
The delivered opening book by Dann Corbit lies in the doc folder.
Zeta supports Polyglot .bin books, just name the file "book.bin" and put it in
the same directory like the engine.

### Compile from source ###
The source code and a GNU/Linux Makefile are placed in src folder.

### Thanks To ###
- All CCC members, http://www.talkchess.com/
- CPW Team, http://chessprogramming.wikispaces.com/
- Dann Corbit, for his help in many cases.
- Gerd Isenberg, for the Quad-Bitboard idea,
  http://chessprogramming.wikispaces.com/search/view/Quad-Bitboard
- Harm Gert Muller, for publishing such a detailed documentation of MicroMax.
- Jim Ablett, for commpiling all the Zeta versions.
- Marcel van Kervinck, for his smart Chess Programm MSCP.
- Michel Van den Bergh for the Polyglot probing code.
  http://hardy.uhasselt.be/Toga/book_format.html
- Steffan Westcott for the KoggeStone move generator,
  http://chessprogramming.wikispaces.com/Kogge-Stone+Algorithm
- Stockfish Team for their great engine.
- Tomasz Michniewski, for his Piece Square Tables,
  https://chessprogramming.wikispaces.com/Simplified+evaluation+function



ChangeLog
ZetaDva (0301) alpha; urgency=low

  * code cleanup and fixes
  * switch from CECP2 (Winboard/Xboard) to UCI
  * eval fixes
  * time control fixes
  * switch to ANSI C89 style
  * switch to GPL >=v2
  * generalized Kogge-Stone move generator
  * QuadBitboard board presentation

 -- Srdja Matovic <s.matovic@app26.de>  Wed, 13 Jul 2013 13:07:16 +0200

ZetaDva (0220) alpha; urgency=low

  * fix in castle move generation
  * added evaluation terms

 -- Srdja Matovic <s.matovic@app26.de>  Sat, 06 Jul 2013 06:07:13 +0200

ZetaDva (0210) alpha; urgency=low

  * basic engine running
  * support for basic xboard commands
  * alphabeta search with QS,MVV-LVA,IID,TT,Nullwindows,Nullmove,LMR,etc.
  * magic bitboard move generator
  * polyglot book support
  * time control for conventional clock mode
  * simplified evaluation

 -- Srdja Matovic <s.matovic@app26.de>  Tue, 10 Jan 2012 10:01:12 +0100



src/types.h
/*
  Name:         Zeta Dva
  Description:  Amateur level chess engine
  Author:       Srdja Matovic <s.matovic@app26.de>
  Created at:   2011-01-15
  Updated at:   2016-07-13
  License:      GPL >= v2

  Copyright (C) 2011-2016 Srdja Matovic

  Zeta Dva is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 2 of the License, or
  (at your option) any later version.

  Zeta Dva is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
*/

#ifndef TYPES_H_INCLUDED
#define TYPES_H_INCLUDED

/* C89 style typedefs
typedef unsigned char bool;
typedef unsigned char u8;
typedef signed short s16;
typedef signed int s32;
typedef unsigned int u32;
typedef unsigned long long u64;
#define false   0
#define true    1
*/
/* C99 headers */
#include <stdint.h>
#include <stdbool.h>
/* C99 types */
typedef uint8_t         u8;
typedef int16_t         s16;
typedef int32_t         s32;
typedef uint32_t        u32;
typedef uint64_t        u64;
/* custom typedefs */
typedef s16             Score;
typedef u64             Bitboard;
typedef u64             Hash;
typedef u64             Cr;
typedef u64             Hmc;
typedef u64             Square;
typedef u64             Piece;
typedef u64             PieceType;
typedef u64             Move;
typedef u64             File;
typedef u64             Rank;

#define VERSION         "0301"

/* quad bitboard array index definition */
#define QBBBLACK  0     /* pieces white */
#define QBBP1     1     /* piece type first bit */
#define QBBP2     2     /* piece type second bit */
#define QBBP3     3     /* piece type third bit */
#define QBBPMVD   4     /* piece moved flags, for castle rights */
#define QBBHASH   5     /* 64 bit board Zobrist hash */
#define QBBLAST   6     /* lastmove + ep target + halfmove clock + move score */
/* move encoding 

   0  -  5  square from
   6  - 11  square to
  12  - 17  square capture
  18  - 21  piece from
  22  - 25  piece to
  26  - 29  piece capture
  30  - 35  square en passant target
  36  - 43  halfmove clock for fity move rule, last capture/castle/pawn move
  44  - 47  4 bit free
  48  - 63  move score
*/
/* engine defaults */
#define MAXPLY      128     /* max internal search ply */
#define MAXGAMEPLY  1024    /* max ply a game can reach */
#define MAXMOVES    256     /* max amount of legal moves per position */
/* colors */
#define BLACK               1
#define WHITE               0
/* scores */
#define INF                 32000
#define MATESCORE           30000
#define DRAWSCORE           0
#define STALEMATESCORE      0
/* piece type enumeration */
#define PNONE               0
#define PAWN                1
#define KNIGHT              2
#define KING                3
#define BISHOP              4
#define ROOK                5
#define QUEEN               6
/* bitboard masks, computation prefered over lookup */
#define SETMASKBB(sq)       (1ULL<<(sq))
#define CLRMASKBB(sq)       (~(1ULL<<(sq)))
#define BBEMPTY             0x0000000000000000ULL
#define BBFULL              0xFFFFFFFFFFFFFFFFULL
#define MOVENONE            0x0000000000000000ULL
#define HASHNONE            0x0000000000000000ULL
#define CRNONE              0x0000000000000000ULL
#define SCORENONE           0x0000000000000000ULL
/* set masks */
#define SMMOVE              0x000000003FFFFFFFULL
#define SMSQEP              0x0000000FC0000000ULL
#define SMHMC               0x00000FF000000000ULL
#define SMCRALL             0x8900000000000091ULL
#define SMSCORE             0xFFFF000000000000ULL
/* clear masks */
#define CMMOVE              0xFFFFFFFFC0000000ULL
#define CMSQEP              0xFFFFFFF03FFFFFFFULL
#define CMHMC               0xFFFFF00FFFFFFFFFULL
#define CMCRALL             0x76FFFFFFFFFFFF6EULL
#define CMSCORE             0x0000FFFFFFFFFFFFULL
/* castle right set masks big endian */
#define SMCRWHITE           0x0000000000000091ULL
#define SMCRWHITEQ          0x0000000000000011ULL
#define SMCRWHITEK          0x0000000000000090ULL
#define SMCRBLACK           0x8900000000000000ULL
#define SMCRBLACKQ          0x8800000000000000ULL
#define SMCRBLACKK          0x0900000000000000ULL
/* move helpers */
#define MAKEPIECE(p,c)     (((p)<<1)|c)
#define GETCOLOR(p)        ((p)&0x1)
#define GETPTYPE(p)        (((p)>>1)&0x7)      /* 3 bit piece type encoding */
#define GETSQFROM(mv)      ((mv)&0x3F)         /* 6 bit square */
#define GETSQTO(mv)        (((mv)>>6)&0x3F)    /* 6 bit square */
#define GETSQCPT(mv)       (((mv)>>12)&0x3F)   /* 6 bit square */
#define GETPFROM(mv)       (((mv)>>18)&0xF)    /* 4 bit piece encoding */
#define GETPTO(mv)         (((mv)>>22)&0xF)    /* 4 bit piece encoding */
#define GETPCPT(mv)        (((mv)>>26)&0xF)    /* 4 bit piece encodinge */
#define GETSQEP(mv)        (((mv)>>30)&0x3F)   /* 6 bit square */
#define SETSQEP(mv,sq)     (((mv)&CMSQEP)|(((sq)&0x3F)<<30))
#define GETHMC(mv)         (((mv)>>36)&0xFF)   /* 8 bit halfmove clock */
#define SETHMC(mv,hmc)     (((mv)&CMHMC)|(((hmc)&0xFF)<<36))
#define GETSCORE(mv)       (((mv)>>48)&0xFFFF) /* signed 16 bit score */
#define SETSCORE(mv,score) (((mv)&CMSCORE)|(((score)&0xFFFF)<<48)) 
/* pack move into 64 bits */
#define MAKEMOVE(sqfrom, sqto, sqcpt, pfrom, pto, pcpt, sqep, hmc, score) \
( \
     sqfrom      | (sqto<<6)  | (sqcpt<<12) \
  | (pfrom<<18)  | (pto<<22)  | (pcpt<<26) \
  | (sqep<<30)   | (hmc<<36)  | (score<<48) \
)
/* square helpers */
#define MAKESQ(file,rank)   ((rank)<<3|(file))
#define GETRANK(sq)         ((sq)>>3)
#define GETFILE(sq)         ((sq)&7)
#define GETRRANK(sq,color)  ((color)?(((sq)>>3)^7):((sq)>>3))
#define FLIP(sq)            (((sq)^7))
#define FLOP(sqe)           (((sq))^56)
/* piece helpers */
#define GETPIECE(board,sq)  ( \
                               ((board[0]>>(sq))&0x1)\
                           |  (((board[1]>>(sq))&0x1)<<1) \
                           |  (((board[2]>>(sq))&0x1)<<2) \
                           |  (((board[3]>>(sq))&0x1)<<3) \
                             )
#define GETPIECETYPE(board,sq) ( \
                           |  (((board[1]>>(sq))&0x1)) \
                           |  (((board[2]>>(sq))&0x1)<<1) \
                           |  (((board[3]>>(sq))&0x1)<<2) \
                             )
/* file enumeration */
enum Files
{
  FILE_A, FILE_B, FILE_C, FILE_D, FILE_E, FILE_F, FILE_G, FILE_H, FILE_NONE
};
#define BBNOTHFILE          0x7F7F7F7F7F7F7F7F
#define BBNOTAFILE          0xFEFEFEFEFEFEFEFE
/* rank enumeration */
enum Ranks
{
  RANK_1, RANK_2, RANK_3, RANK_4, RANK_5, RANK_6, RANK_7, RANK_8, RANK_NONE
};
/* square enumeration */
enum Squares
{
  SQ_A1, SQ_B1, SQ_C1, SQ_D1, SQ_E1, SQ_F1, SQ_G1, SQ_H1,
  SQ_A2, SQ_B2, SQ_C2, SQ_D2, SQ_E2, SQ_F2, SQ_G2, SQ_H2,
  SQ_A3, SQ_B3, SQ_C3, SQ_D3, SQ_E3, SQ_F3, SQ_G3, SQ_H3,
  SQ_A4, SQ_B4, SQ_C4, SQ_D4, SQ_E4, SQ_F4, SQ_G4, SQ_H4,
  SQ_A5, SQ_B5, SQ_C5, SQ_D5, SQ_E5, SQ_F5, SQ_G5, SQ_H5,
  SQ_A6, SQ_B6, SQ_C6, SQ_D6, SQ_E6, SQ_F6, SQ_G6, SQ_H6,
  SQ_A7, SQ_B7, SQ_C7, SQ_D7, SQ_E7, SQ_F7, SQ_G7, SQ_H7,
  SQ_A8, SQ_B8, SQ_C8, SQ_D8, SQ_E8, SQ_F8, SQ_G8, SQ_H8
};

#endif /* TYPES_H_INCLUDED */

